pandas/_libs/src/ujson/python/date_conversions.c

```diff
@@ -79,7 +79,9 @@ char *PyDateTimeToIso(PyObject *obj, NPY_DATETIMEUNIT base,
     npy_datetimestruct dts;
     int ret;
 
-    ret = convert_pydatetime_to_datetimestruct(obj, &dts);
+    PyDateTime_IMPORT;
+
+    ret = convert_pydatetime_to_datetimestruct((PyDateTime_Date*)obj, &dts);
```

Comment:
Make sure to avoid blind casting - these should all be preceded by a `PyDate_Check` call with appropriate error handling, otherwise its anybody guess what happens when you pass the wrong object type to this

Reply:
so something like

```
    PyDateTime_IMPORT;
    if (!PyDate_Check(obj)) {
        PyErr_SetString(PyExc_ValueError,
                        "Object is not a datetime.date object");
        return NULL;
    }
    PyDateTime_Date *dtobj = (PyDateTime_Date*)obj;

    ret = convert_pydatetime_to_datetimestruct(dtobj, &dts);
```

Reply:
So in objToJSON.c L1552 we call PyDateTimeToEpoch but IIUC `obj` ther eis a np.datetime64 object, not a pydatetime object.  Isn't that call doomed to failure?

Reply:
Yea you have got the right idea with that error handling - nice job

Reply:
If that is the case I think that behavior would be undefined, so could segfault or get garbage values. Do you know of a reproducer?